src/main/java/jnr/ffi/provider/jffi/DirectMemoryIO.java

```diff
@@ -186,8 +187,30 @@ public String getString(long offset) {
 
 
     public String getString(long offset, int maxLength, Charset cs) {
-        final byte[] bytes = IO.getZeroTerminatedByteArray(address() + offset, maxLength);
-        return cs.decode(ByteBuffer.wrap(bytes)).toString();
+        if(cs == StandardCharsets.UTF_8){
+            final byte[] bytes = IO.getZeroTerminatedByteArray(address() + offset, maxLength);
+            return cs.decode(ByteBuffer.wrap(bytes)).toString();
+        }else{
+            byte[] bytes = new byte[maxLength];
+            IO.getByteArray(address() + offset, bytes, 0, maxLength);
+            final byte[] nullCharBytes = new String("\0").getBytes(cs);
```

Comment:
This could possibly be cached for common encodings (UTF-16, UTF-32) or it *might* be valid to just use a width of `\0` equal to the encoding's CharsetEncoder.maxBytesPerChar value.